dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/JavaArchive.java

```diff
@@ -0,0 +1,575 @@
+/*******************************************************************************
+ * Copyright (c) 2020 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************
+ *
+ * The purpose of this class is to test if a Java archive is valid without actually reading the entire JAR.
+ * This is especially useful in cases where the archive has a script attached at the front of the file.
+ * Of course, this is more of a sanity check than a rigorous validation, which would require reading the whole file.
+ * 
+ *     [ script ]                   <---- start of file  (optional)
+ *     [ local file header 1 ]      <---- actual start of archive
+ *     [ encryption header 1 ]
+ *     [ file data 1 ]
+ *     [ data descriptor 1 ]
+ *     . 
+ *     .
+ *     .
+ *     [ local file header n ]
+ *     [ encryption header n ]
+ *     [ file data n ]
+ *     [ data descriptor n ]
+ *     [ archive decryption header ] 
+ *     [ archive extra data record ] 
+ *     [ central directory header 1 ]       <---- start of central directory
+ *     .
+ *     .
+ *     .
+ *     [ central directory header n ]
+ *     [ zip64 end of central directory record ]
+ *     [ zip64 end of central directory locator ] 
+ *     [ end of central directory record ]
+ * 
+ *  The validation calculates the size of the last 4 records listed above (as well as doing some
+ *  validation checks on the sizes and signatures) and then calculates the actual offset
+ *  of the archive within the file (which will normally be 0).  The final step is to check
+ *  the signature at the beginning of the archive.
+ *  
+ *  The actual archive offset is calculated :
+ *  
+ *    fileLength - eocdRecordSize - zip64RecLengths - CentralDirectorySize - offsetOfCentralDirectory
+ *   
+ *   Note the offsetOfCentralDirectory is relative to the beginning of the actual archive (not an offset  
+ *   from the beginning of the file.)  Subtracting away all of the known pieces gives us the unknown 
+ *   piece which is the actual archive offset.
+ *  
+ */
+
+package com.ibm.ws.artifact.zip.internal;
+
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.util.Arrays;
+
+public class JavaArchive {
```

Comment:
Not sure about using the name "JavaArchive': The target archive is a PK archive, and can be a plain ZIP file, which has nothing to do with java.

Reply:
changed to ZipValidator